idea/src/org/jetbrains/kotlin/idea/inspections/collections/ConvertCallChainIntoSequenceInspection.kt

```diff
@@ -0,0 +1,164 @@
+        val lastCall = calls.lastOrNull() ?: return
+        val first = firstCall.parent as? KtQualifiedExpression ?: return
+        val last = lastCall.parent as? KtQualifiedExpression ?: return
+        val next = last.parent as? KtQualifiedExpression
+        val nextType = next?.getCallableDescriptor()?.returnType
+
+        val psiFactory = KtPsiFactory(expression)
+        val dot = buildString {
+            if (first.receiverExpression.siblings().filterIsInstance<PsiWhiteSpace>().any { it.textContains('\n') }) append("\n")
+            if (first is KtSafeQualifiedExpression) append("?")
+            append(".")
+        }
+
+        val firstCommentSaver = CommentSaver(first)
+        val firstReplaced = first.replaced(
+            psiFactory.buildExpression {
+                appendExpression(first.receiverExpression)
+                appendFixedText(dot)
+                appendExpression(psiFactory.createExpression("asSequence()"))
+                appendFixedText(dot)
+                appendExpression(firstCall)
+            }
+        )
+        firstCommentSaver.restore(firstReplaced)
+
+        val nextDescriptor = next?.getCallableDescriptor()
+        if (nextDescriptor == null
+            || !nextDescriptor.fqNameSafe.asString().startsWith("kotlin.sequences.")
+            || nextDescriptor.returnType != nextType
+        ) {
+            val lastCommentSaver = CommentSaver(last)
+            val lastReplaced = last.replace(
+                psiFactory.buildExpression {
+                    appendExpression(last)
+                    appendFixedText(dot)
+                    appendExpression(psiFactory.createExpression("toList()"))
+                }
+            )
+            lastCommentSaver.restore(lastReplaced)
+        }
+    }
+}
+
+private fun KtQualifiedExpression.findTarget(): KtQualifiedExpression? {
+    if (parent is KtQualifiedExpression) return null
+
+    val calls = collectCallExpression()
+    if (calls.isEmpty()) return null
+
+    val qualified = calls.first().parent as? KtQualifiedExpression ?: return null
+    val fqName = qualified.getCallableDescriptor()?.returnType?.constructor?.declarationDescriptor?.fqNameSafe
+    if (fqName != KotlinBuiltIns.FQ_NAMES.list) return null
+    return qualified
+}
+
+private fun KtQualifiedExpression.collectCallExpression(): List<KtCallExpression> {
+    val calls = mutableListOf<KtCallExpression>()
+
+    fun collect(qualified: KtQualifiedExpression) {
+        val call = qualified.callExpression ?: return
+        calls.add(call)
+        val receiver = qualified.receiverExpression
+        if (receiver is KtQualifiedExpression) collect(receiver)
+    }
+    collect(this)
+
+    if (calls.size < 2) return emptyList()
+
+    val transformationCalls = calls
+        .asSequence()
+        .dropWhile { it !in transformations }
+        .takeWhile { call ->
+            call in transformations && call.valueArguments.none { arg ->
```

Comment:
It's definitely not the fastest way to find the compatible transformation, if any. You have O(n) in your solution, but you can easily get O(1) instead.